### DISCUSSION TITLE
USB host mode possible?

### DISCUSSION THREAD

**Discussion by hippi-viking** [frntc/Sidekick64#42] [Ideas]
Hello,

Circle 44.5 revised (and expanded) its USB support. Do you think it would be possible to operate the Sidekick64 in USB host mode to connect game/MIDI controllers, storage devices to be used on the C64?

**Comment by frntc** [comment]
Hi, 

I haven't tried out Circle 44.5 yet, but the summary of changes says "the FIQ cannot be used for other purpose than the USB with this system option". The problem here is that all communication between the RPi and the 8-bit machine uses FIQs :-( I might have a look at the USB-FIQ handler and see if it can be called from my own FIQs (some polling mode), but no promises when/if this will work. There are alternative USB-drivers (from outside of Circle), but I haven't had a chance to have a closer look.

From a more general point: the problem with USB and SK64 (so far) is that the USB-driver uses an interrupt which interferes with FIQ-timing and caching. On the other hand, USB-game controllers would require special code on the C64 side anyway, which could be prepared for that.

Storage device: what do you have in mind?

**Reply by hippi-viking** [reply]
Hey,

Thanks for getting back! Ah, i see the bottleneck now. Thanks for having a look and thinking about it!

My main use case would be using the C64 as a MIDI synth at the moment, that could work if the RPi could be the host for an USB MIDI controller - but it requires significant effort to do.

For storage: i was thinking that it could be possible to attach and use a flashdrive or a USB-HDD to the RPi much the same way as the SD card.

**Reply by frntc** [reply]
well, MIDI is not the most challenging communication, I'd have to think about it, but maybe one could abuse some GPIOs :-)
As an alternative: have you seen SIDKick?

Storage: if it's an option you could mirror the USB-storage to the SD-card, so that you don't have to access those at runtime. Not elegant though...

**Reply by hippi-viking** [reply]
MIDI: thanks for looking into this! Yes, I had a look at SIDKick, which looks mighty interesting! :) The thing is though that the C64 I am (ab)using now was my late uncle's and I would love to keep it in factory condition for as long as possible. (But once something breaks I will have to put in replacement parts anyway...)

Storage: it was just an idea, probably does not make a lot of sense other than separating the RasPI's things from the Commodore's.
